slowapi/wrappers.py

```diff
@@ -74,13 +75,21 @@ def __init__(
         self.error_message = error_message
         self.exempt_when = exempt_when
         self.override_defaults = override_defaults
+        self.request = None
 
     def __iter__(self) -> Iterator[Limit]:
-        limit_items: List[RateLimitItem] = parse_many(
-            self.__limit_provider()
-            if callable(self.__limit_provider)
-            else self.__limit_provider
-        )
+        if callable(self.__limit_provider):
+            if "key" in inspect.signature(self.__limit_provider).parameters.keys():
+                assert (
+                    "request" in inspect.signature(self.key_function).parameters.keys()
+                )
+                assert self.request
```

Comment:
Good question. I understand `assert` as being primarily a tool to guide coders during development, and help catch coding errors. So if this error will only happen because the code is invalid, I'd keep the `assert` and add a clear error message, otherwise, I'd agree with you and replace both `assert`s with `if` tests.
My impression here is that the first one is probably checking for coding errors, but maybe the second one might need to be something like `if self.request is None:...`
(I've seen this used extremely well in https://github.com/encode/django-rest-framework which gives very clear hints as to what is wrong with your code, and saves a ton of time when developing. I think it's a very good example to follow)

Reply:
Good point that asserts are removed in optimized mode, I didn't know that. However, I agree with @laurentS that it's nice way to guide coders, so I kept assert and added error message.
I replaced the second check to `if self.request is None: raise Exception(...)`. But I noticed that in `async_wrapper` and `sync_wrapper` there's already a check for correct `request`:
```
if not isinstance(request, Request):
    raise Exception(
        "parameter `request` must be an instance of starlette.requests.Request"
    )
```
So the second assert is never fired and maybe I should just remove it.